---
title: "Example code for single-cell analysis with Seurat, day 2"
author: "Devika Agarwal/updated by Carla Cohen"
date: "23/05/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(tidyverse)
library(patchwork)
library(DT)
library(gprofiler2)

```

# Exercise

## Read in the RDS object we created and save from Seurat day 1

- use the `readRDS()` function to read in previously saved object

```{r}
seurat_after_qc <- readRDS("Seurat_after_QC.Rds")
```

- Check the Default Assay

```{r}
DefaultAssay(seurat_after_qc)
```


## Apply SCTransfrom  normalisation

Use `SCTransform()` function

SCTransform vignette: https://satijalab.org/seurat/articles/sctransform_vignette.html 

`SCTransform()`  command replaces `NormalizeData()`, `ScaleData` and `FindVariableFeatures()` run for the RNA assay in day 1 Seurat

Should we remove any confounding variables like we did for the RNA assay for Day 1?

Do we want to use the same number of variable featuresn(n=1000) or more than what we used for `NormalizeData()` function.


```{r}
seurat_after_qc <- SCTransform(seurat_after_qc,
                               vars.to.regress = c("percent_mt")) 

#do.correct.umi will correct for sequencing depth (leave as true)
#ncells is the number of subsampling cells used to build the model, default is 5000
#since the modeling it better it can identify more variable genes, so the default is 3000 for variable.features.n.
#we could change it with variable.features.n = 2000
#vars.to.regress - we don't need to regress UMI because this is included in the transform already so you would be doing it twice.
#it sets a seed by default
#return.only.var.genes is TRUE to save memory. You could change it to false if you wanted the scaled data matrix for all genes.
#
```
where is the new normalisation stored?
Answer: It is an assay called SCT, this has been set as the default assay.

explore the `seurat_after_qc` objects meta.data and assays ? 


```{r}
Assays(seurat_after_qc) 
seurat_after_qc[[]] 

```

is there a change? 
Answer: yes, new Assay called SCT 

Are there new columns in the meta.data?
Answer: new columns nCount_SCT and nFeature_SCT

Type seurat_after_qc in the console to get the key info.

```{r}
dim(seurat_after_qc@assays$RNA) #19037 4204
dim(seurat_after_qc@assays$SCT) #17463 4204

#We have got fewer genes in the SCT assay but same number of cells

?sctransform::vst #vst is the package under the hood of sctransform
#we see that the mimimum number of cells expressing a gene is 5, whereas in log normalization transform it is 3. 
#therefore we have fewer cells output in the SCT assay output

?SCTransform
#counts are the corrected counts
#data is log1p(counts)
#scale.data is pearson residuals

```


# Exercise

## Visulaisation

- The library size and number of features detected per cell is already present in the Seurat object.

- When you run `SCTransform` you get two new variables for library size and features for SCT normalisation

 - Use the function `VlnPlot()` to display compare RNA assay vs SCT assay library size in one plot 
 
 - Vlnplot of features detected in RNA vs SCT  in a single violin plot.

- check to see how you can have fixed ylims in the `VlnPlot()` function

```{r}

colnames(seurat_after_qc[[]])

VlnPlot(seurat_after_qc,
        features = c("nCount_RNA", "nCount_SCT"),
        pt.size = 0, #get rid of the dots
        same.y.lims = TRUE) #same y axis on both plots

```
We can see there is a big variation in sequencin depth across the clusters.
This is normalised during the SCTransform. 
There is still some variation in the normalised data, but there is still variation (because there are indeed some genes that are highly expressed). 
SCT is getting rid of the technical effect while leaving the biological signal.



## Bonus-
- Visualise the distribution of SCT corrected count data for a feature of your choice.
- Visualise the distribution of SCT logs counts in the data slot for a feature (e.g. LYZ)

- Visualise the distribution of SCT scale.data count feature of your choice. 

lets choose LYZ like day 1

```{r}
DefaultAssay(seurat_after_qc) <- "SCT" #set the correct assay because we cannot specify the assay in FetchData
FetchData(seurat_after_qc, vars = "LYZ", slot = "counts")
#3 slots available (described in the SCTransform help section)
#"counts", corrected for sequencing depth
#"data", log normalised of the UMI corrected counts. It is more sparse.
#"scale.data" is the fully SCTransformed data, this is where the real signal is seen. It looks similar to the log norm data from yesterday
#we can choose which of these to use in the downstream analysis
#here we are plotting all three so that we can easily compare them.

ggplot_lyz_corrected <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "counts"), aes(LYZ)) +
    geom_histogram(fill = "grey", color = "black", bins = 100) +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()
ggplot_lyz_normalised <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "data"), aes(LYZ)) +
    geom_histogram(fill = "grey", color = "black", bins = 100) +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()
ggplot_lyz_scaled <- ggplot(FetchData(seurat_after_qc, vars = "LYZ", slot = "scale.data"), aes(LYZ)) +
    geom_histogram(fill = "grey", color = "black", bins = 100) +
    coord_cartesian(ylim = c(0, 500)) +
    cowplot::theme_cowplot()


cowplot::plot_grid(ggplot_lyz_corrected, ggplot_lyz_normalised,ggplot_lyz_scaled, ncol = 1)
```
Use the function `VariableFeaturs` to pull out the 1:10 the variable genes after SCT and compare to 1:10 from the RNA assay

do we need to change any arguments to get the variables genes specific to the SCT or RNA assay

```{r}
VariableFeatures(seurat_after_qc, assay = "RNA")[1:10]
VariableFeatures(seurat_after_qc, assay = "SCT")[1:10]
```

We can see there there are some common genes but the order is not exactly the same.
e.g. LYZ is the most highly variable gene in the transfromed data.
You could check how many variable genes are similar.


# Exercise


## Dimensionality reduction on SCT transformed data

- Run a principal component analysis and UMAP on the Seurat object.
- Check the Default assay

- Do we want to change the `reduction.name` argument so that we can still keep the RNA assay based PCA results?
  
```{r}
DefaultAssay(seurat_after_qc) <- "SCT"
seurat_after_qc <- RunPCA(seurat_after_qc, reduction.name = "sct.pca")
```
We currently have pca and umap from our analysis yesterday. These must be on the RNA assay.
After running the above command we now have a new reduction called sct.pca. 
If we had not renamed then it would have over-written the RNA PCA from yesterday.


check to see what reductions are now present in the object

```{r}
Reductions(seurat_after_qc)
```

- First, visualise the amount of variance explained the top principal components for SCTransformed data (number of your choice).
  How many principal components would you use for downstream analyses?
  
- do we need to specificy the correction reduction?

```{r}
#we should specify which reduction
ElbowPlot(seurat_after_qc, ndims = 50, reduction = "sct.pca")
```

- Run the UMAP technique on the first 20 principal components stored in the Seurat object and visualise the result as a scatterplot.

how can we change the reduction name from defauult "umap" to "sct.umap"

how can we specify that we want to use pca run on the SCT Assay (sct.pca) in the previous step?



```{r}
seurat_after_qc <- RunUMAP(seurat_after_qc,
                           reduction = "sct.pca",
                           dims = 1:20,
                           reduction.name = "sct.umap")
```

use DimPlot() to plot the umap, what happens if you try to specify different reductions with UMAPPlot()?

compare RNA based umap with sct.umap

```{r, fig.width=10}
p1 <- DimPlot(seurat_after_qc, reduction = "umap")  + ggtitle("RNA UMAP")
p2 <- DimPlot(seurat_after_qc, reduction = "sct.umap")  + ggtitle("SCT UMAP")
p1 + p2
```

SCT does a slightly better job of giving you defined clusters.

# Exercise

## Clustering on SCTransformed data

- Then, compute the graph of nearest neighbours using the function `FindNeighbors()`.
  Which principal components are used by default?
  Instead, specify the number of components that you have chosen.
  Have you chosen the right `reduction`  and `assay`? 
  
```{r}
seurat_after_qc <-  FindNeighbors(seurat_after_qc,
                                  reduction = "sct.pca",
                                  dims = 1:20)
```


- Finally, compute cluster labels.
  What is the default setting for the `resolution` argument?
  Instead, set it to `0.5`.
  Do you expect more or fewer clusters following that change?
  What other parameters would you also try to experiment with?
  - should we specify the `graph.name` ?

```{r}
seurat_after_qc <- FindClusters(seurat_after_qc,
                                resolution = 0.5,
                                graph.name = "SCT_snn")

Graphs(seurat_after_qc)
```


Check cluster assignment between SCT and RNA workflow
- We get a different number of clusters (13 vs 12)
if you use the same resolution = o.5 and dims as RNA workflow do you get the same number of cluster or more?

are cells in the same cluster across both RNA and SCT
```{r}
table()

```

- Visualise the SCT cluster labels on the SCT transformed UMAP scatter plot and the RNA cluster labels on the RNA umap
  How would you describe the agreement between the UMAP layout and the clustering results?= for SCT vs RNA Assay results
  

```{r}
seurat_after_qc[[]]

p1 <- DimPlot(seurat_after_qc, reduction = "umap", label = TRUE, group.by = "RNA_snn_res.0.5") + ggtitle("RNA UMAP")
p2 <- DimPlot(seurat_after_qc, reduction = "sct.umap", label = TRUE, group.by = "SCT_snn_res.0.5") + ggtitle("SCT UMAP")
p1 + p2

```


Plot some known cell-type markers for PBMC datasets, does the SCT better seperate the celltypes?

CD14+ Monocyte : LYZ, CD14
CD16 Monocytes : FCGR3A, MS4A7

CD4 T :  CD4, IL76

CD8 T : CD8A, CD3D

NK : GNLY, GZMB,NKG7

B Cell :  MS4A1 , CD79A

DC : CST3, FCER1A

Platelets : PPBP


```{r, fig.width=10, fig.height=8}
FeaturePlot(seurat_after_qc, features = c("LYZ","FCGR3A","CD4","CD8A","GNLY","MS4A1","CST3","PPBP"), reduction = "sct.umap")
FeaturePlot(seurat_after_qc, features = c("LYZ","FCGR3A","CD4","CD8A","GNLY","MS4A1","CST3","PPBP"), reduction = "umap")
```

We can see how the CD8 T cells separate better with the SCT.
We could also try adding more PCs so see if we get better clustering.

# Bonus exercise to try in your own time:: Pathway analysis on Cluster markers for all clusters


- Choose either RNA workflow based generated or SCT generated based  seurat marker results

- we will be using Gprofiler `gost()` function fot multiple gene lists at the same time

- First we need to  filter seurat get top significant (p_adj < 0.05)  upregulated genes with a LogFC threshold (decided by you) for each cluster , use `dplyr:: group_by()` and `dplyr::filter` to get gene list for each cluster
then only select cluster and gene

- We then use ` split()`  on the filtered_df to divide gene markers into list  of multiple character vectors containing genes split by cluster

- can refer to the pahtway analysis code from week 3 but for h.sapiens

```{r}
seurat_clusters_results_filtered <- seurat_markers_all_sct %>% dplyr::group_by() %>% filter() %>%select(cluster,gene)
seurat_clusters_list <- split(seurat_clusters_results_filtered,cluster)
```


 We then run pathway analysis using `gost()` with multi_query = TRUE
 
 - to generate all_genes_id , we use all genes present in either the RNA assay or SCT assay , as we have already filtered out lowly expressed genes not present in certain cells .
 
```{r}
# Choose Default assay based on if running pathway analyses on RNA or SCT results
DefaultAssay(seurat_after_qc) <- ""
# create a vector of  of all genes 
all_genes_id <- rownames(seurat_after_qc)

multi_gostquery_results_obj <- gost(query = seurat_clusters_list,
                                    organism = "hsapiens", 
                                    exclude_iea = TRUE, # change to True as have alot of results otherwise
                                    correction_method = "g_SCS",  # this is thier own multiple testing coorrection - actually more strignent then bonferroni or BH 
                                    custom_bg = all_gene_ids, # make sure this is all gene ids 
                                    multi_query = TRUE)
```


can you plot the results for different clusters together ?

```{r}

gostplot(multi_gostquery_results_obj, capped = TRUE, interactive = TRUE)
```



Afternoon Session

Demultiplexing with hashtag oligos (HTOs)

Dataset : 12-HTO dataset from four human cell lines

* Dataset represent Data represent single cells collected from four cell lines: HEK, K562, KG1 and THP1
* Each cell line was further split into three samples (12 samples in total)
* Each sample was labeled with a hashing antibody mixture (CD29 and CD45), pooled, and run on a single lane of 10X.
* Based on this design, we should be able to detect doublets both across and within cell types


Load in the UMI matrix for the RNA data
```{r}
hto12.umis <- readRDS("/project/obds/shared/resources/4_r_single_cell/seurat_day_2/data/hto12_umi_mtx.rds")

# how many cells and genes do we have in the RNA matrix?

dim(hto12.umis) #25339 (genes) x 30000 (cells)
head(hto12.umis)[1:5,1:5]

#rownames are genes
#columns are cells
#this is the same as for the seurat object.
```


load in the HTO matrix
This gives you the antibody tags information

```{r}
hto12.htos <- readRDS("/project/obds/shared/resources/4_r_single_cell/seurat_day_2/data/hto12_hto_mtx.rds")

# what is the structure of the hto12.htos object? (what is the rows and what is the columns?)
str(hto12.htos)
# check the class of the hto object
class(hto12.htos) #it's a df
head(hto12.htos)
#how many cell barcodes and HTO samples do we have?
dim(hto12.htos) #687242 x 16

#rownames are cells, columns are sample type
```
Rownames are cell barcodes.


Now we only want to subset to those cell barcodes or  cells (actually called as cells by cellRanger or EmptyDrops on the gene expression data) which are detected by both RNA and HTO matrices? 

```{r}
#we need row names from hto and col names from UMIs
head(rownames(hto12.htos))
head(colnames(hto12.umis))

joint.bcs <- intersect(rownames(hto12.htos), colnames(hto12.umis))

# check the class  of the joint.bcs object and how many cell barcodes do we have in common
class(joint.bcs) #character vector

length(joint.bcs) #25088

```
So there are 25088 barcodes in common between the two matrices. 

In the HTO we only have the sequence of the barcode. 


# subset  the RNA matrix  to only the `joint.bcs` cell barcodes

```{r}

head(joint.bcs)

hto12.umis.common <- hto12.umis[,joint.bcs]

# check the dims
dim(hto12.umis.common) #25339 25088
class(hto12.umis.common) #matrix
head(hto12.umis.common)[1:5, 1:5]

```




- Create  a Seurat object with the RNA (UMI count matrix) only containing the `joint_bcs`

 Name the object `hto12_object`
 
 Include features detected in at least 3 cells,
 
 and cells where at least 200 features detected
 
- add HTO data as another assay to `hto12_object` 




```{r}
#Create seurat object
hto12_object <- CreateSeuratObject(counts = hto12.umis.common,
                                   min.cells = 3,
                                   min.features = 200)


# Normalise with log normalization ,find variable genes and Scale RNA data
hto12_object <- NormalizeData(hto12_object)
hto12_object <- FindVariableFeatures(hto12_object)
hto12_object <- ScaleData(hto12_object)

dim(hto12_object@assays$RNA@scale.data) #the scaled data is on the 2000 variable genes
```



we only want to subset the hto matrix to those cell barcodes which are now in the `hto12_object` Seurat object and  make sure that the features only consist of the HTO tags

Now look at the function `CreateAssayObject()` to  add the subsetted  HTO  matrix to the already created  `hto12_object` seurat object as a new assay called `HTO`

what is the expected format of the input matrix for `CreateAssayObject()`?

answer: 

is our subsetted `hto12.htos.common` in the right format? if not what do we do to get it in the right format before adding it as another assay?

Answer : 


```{r}
# we want to subset the hto12.htos object such that  the rows or cell barcodes are only the same ones as the hto12_object and colnames are only consist of the 12 HTO tags
# we then want to transpoose this matrix, so now our cell barcodes are colnames and HTO tags are rownames 

head(hto12.htos) # cells are row names
colnames(hto12_object) #this is what we are adjusting


hto12.htos.common <- t(hto12.htos[colnames(hto12_object), 1:12])
#subset using [] the htos that match the names in hto12_object
#select the first 12 rows that are relevant
#transpose the matrix so that it has row names as cell tags

```



do we want to do any further filtering on the HTO object?

Answer : No because we filter on UMIs not HTOs

now add a new assay called "HTO" to the `hto12_object` seurat object

```{r}

hto12_object[["HTO"]] <- CreateAssayObject(counts = hto12.htos.common)

#check the dim of the seurat object
dim(hto12_object)
```


Normalise the HTO data , here we will use the CLR transformation with margin =1 (Default setting)

Here we are finding out the relative expression of each gene per each sample type (?)

```{r}
# check the Default Assay
DefaultAssay(hto12_object) #RNA

hto12_object <- NormalizeData(hto12_object, assay = "HTO", normalization.method = "CLR" , margin=1)
```


Demultiplex cells based on HTO enrichment

Here we use Seurat Function `HTODemux()` to assign single cells to their original samples

```{r}
hto12_object <- HTODemux(hto12_object)
```

checkout the metadata column of the `hto12_object` , try to read the `HTODemux()` results output summary in the `Value` section to understand the results

```{r}
hto12_object[[]]

```
It looks for the highest and second highest likely sample type for each cell. 
Then it uses this information to classify it as one of the samples, or as a doublet if the top two are different. 
The final useful column is hash.ID which gives you the final output. 


Visualise the Demultiplexing results

we can visualise how many cells are classified as singlets, doublets and negative/ambiguous cells

check the meta.data, which column do we want for this information? HTO_classification.global


```{r}
table(hto12_object$HTO_classification.global)

```

Visualize enrichment for selected HTOs with ridge plots

plot the max HTO signal  for one of the HTO of each of the 4 cell lines (HEK, K562, KG1 and THP1) features with ridge plots using the `RidgePlot()` function

plot Max HTO signal

```{r, fig.height=8, fig.width=10}
# Change the identities of the seurat object to the relevant metadata column

Idents(hto12_object) <- "HTO_maxID"
RidgePlot(hto12_object, 
          features = c("HEK-A", "THP1-A", "K562-A", "KG1-A"),
          assay = "HTO",
          ncol = 2)
```



Visualize pairs of HTO signals to confirm mutual exclusivity in singlets  between the same celline

a) plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by (single/doublet/negative status) 

b) plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by HTO_maxID 

c) plot scatter plot of 2 HTOs within the same cell line e.g. HEK, colour by HTO_secondID

use the function `FeatureScatter()`

```{r}
DefaultAssay(hto12_object) <- "HTO"

FeatureScatter(hto12_object,
               feature1 = "HEK-A",
               feature2 = "HEK-B",
               group.by = "HTO_classification.global",
               span = F, smooth = F, shuffle = T)

FeatureScatter(hto12_object,
               feature1 = "HEK-A",
               feature2 = "HEK-B",
               group.by = "HTO_maxID",
               span = F, smooth = F, shuffle = T)

FeatureScatter(hto12_object,
               feature1 = "HEK-A",
               feature2 = "HEK-B",
               group.by = "HTO_secondID",
               span = F, smooth = F, shuffle = T)



```
what do you notice ?

1) We have some doublets
2) The doublets are composed of HEK-A and HEK-B



Bonus Exercise

Plot scatter plot of 2 HTOs across different cell lines e.g. K562 vs KG1 and colour by (single/doublet/negative status) and HTO_max ID
 
```{r}

```
 
 
 
 
 
 
 Compare number of RNA UMIs for singlets, doublets and negative cells
 
 what is a suitable plot for such comparisons?
 
 Answer: 

```{r}
Idents(hto12_object) <- 

```

question: what do you notice?

Answer: 


Visualize HTO signals in a heatmap , lookup `HTOHeatmap()`

```{r, fig.width=8}
HTOHeatmap()
```

what do you notice?



Generate a two dimensional tSNE or umap embedding for HTOs. Here we are grouping cells by singlets and doublets ONLY for simplicity.

do we need to subset our object?

if so what are we subsetting out?




```{r}

Idents(hto12_object) <- 
hto12_object.subset <- subset()
```


Run UMAP/TSNE
 
what assay are we running UMAP/tsne for ?

look up the arguments in `RunUMAP()` and/or `RunTSNE()` functions

check which arguments in RunUMAP/RunUMAP/RunTSNE can be used to change the name of the reduction from defauult name of pca/umap/tsne to custom name

before we Run UMAP, we need to scale and run PCA like we did in the normal single cell workflow


Answer:


```{r}
# Calculate a tSNE & UMAP embedding of the HTO data
DefaultAssay(hto12_object.subset) <- "HTO"




hto12_object.subset <- RunUMAP()

```

check the Reductions in the object

```{r}
Reductions()
```


Plot the UMAP/tsne for the HTO assay
* which reduction shall we plot?

• colour by if singlet/doublet

• colour by HTO final classification results (hash.ID)

* check the arguments on how label the clusters by the cluster identity

* can chnage the label size?


```{r}

```
what do you notice about the cluustering on tthe UMAP/tsne, does the number of clusters mean anything?

Answer: 

 
what do you notice about the cloud of cells surrounding each cluster?

Answer: 


Bonus exercise
# You can also visualize the more detailed classification result by group.by 
HTO_maxID before plotting. 

```{r}

```
what happens if you group.by the UMAP/TSNE plot by HTO..maxID?

Answer: 






Bonus Exercise (if we have time)


Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.

do we need to rerun `FindVariableFeatures()` and `ScaleData()` again?

Answer : 

what other steps  do we need run to get viuslise our RNA data as umap/t-SNE? coloured by doublets/singlets/ and celltypes

answer: 


```{r}
DefaultAssay(hto12_object.subset) <- "RNA"
# Run PCA on most variable features
hto12_object.subset <- 
hto12_object.subset <- 


hto12_object.subset <- RunPCA(hto12_object.subset)
hto12_object.subset <- RunUMAP(hto12_object.subset, dims = 1:8)

```



Plot RNA based UMAP

group.by hash.ID 

create a new seurat object  meta.data column called _cell_line , which removes "_A or _B or _C " in the hash.ID and replaces it with "", to create a new meta.data with only the cell-line info

```{r}
#we create another metadata column based on the hash.id column, where we gsub the HTO tag info (-A,-B,-C) for each cell line to plot only the cell lien names to see if we have batch effect

hto12_object.subset$cell_line <- gsub(pattern = "[-ABC]")
```


```{r}
DimPlot()
```

what does our RNA based clustering on the UMAP/T-SNE show?

Answer: 
        


Bonus exercise (try in your own time)

create a second seurat object based , using the code above, and rerun the `HTODemux()` with a different value of pasitive quantile.

try to check if the classificatiion changes massively if you adjusted the threshold for classification by playing around with the `positive.quantile` argument from the default.


